### DISCUSSION TITLE
Reading Hill Constants through CSV file

### DISCUSSION THREAD

**Discussion by mvkpenn22** [idaholab/moose#26495] [Q&A Modules: Solid mechanics]
Hi,

I am working with a model in which I am defining Hill Constants for my model. But, I want to define these 6 constants directly from the CSV file 

hill_constants.csv:
0.5 1.0 0.5 1.5 1.5 1.5

[VectorPostprocessors]
  [hill_values]
    type = CSVReader
    csv_file = hill_constants.csv
    header = true
    outputs = none
  []
[]

[Materials]
  [mat]
    type = GenericFunctionMaterial
    prop_names = hill_constants
    prop_values = 'hill values'
  []
  [hill_tensor]
    type = ADHillConstants
    # F G H L M N
    hill_constants = hill_constants
    use_large_rotation = true
  []

However, I am getting an error

*** ERROR ***
/storage/work/mooseApp/hill.i:63.5: cannot convert field 'Materials/hill_tensor/hill_constants' value 'hill_values' to float

**Comment by GiudGiud** [comment]
Hello

The ADHillConstants material is expecting a single number, a float, not a vector from a vector postprocessor like you are doing now.

Is there another class you could use for that model?
If not you ll need to modify it to make it accept the format you want.

Guillaume

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

It is not a number, in the original code, it is defined like this

[hill_tensor]
type = ADHillConstants
# F G H L M N
hill_constants = '0.5 1.0 0.5 1.5 1.5 1.5'
use_large_rotation = true
[]

Instead of this, I defined using vector

**Reply by GiudGiud** [reply]
Agreed it is not a float. It s a vector then. Still not a vector postprocessor (what you get from a CSVReader). The parameter expects a vector that has to be explicitly written down in the input. 

My previous advice still stands.

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,

Do I need to use this

[Functions]
  [block]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_block'
    read_type = 'block'
    column_number = '2'
  []
[]

[UserObjects]
  [reader_block]
    type = PropertyReadFile
    prop_file_name = 'hill_constants.csv'
    read_type = 'block'
    nprop = 5
    nblock = 1
  []
[]

But this is reading single column. For my case, I want to read all columns at a time. Can you please tell me how to do this?

But

**Reply by GiudGiud** [reply]
Hello

This can get you each component of the Hill constants as a function. You would just need to use 6 functions.
Unfortunately this is still not enough. If you look at the docs for https://mooseframework.inl.gov/source/materials/HillConstants.html
you can see it only expects constants. Not functions either. 

Can you look at the other HillXX materials to check whether they work for you? (e.g. they accept either a Vector from VectorPostprocessor or 6 functions)

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

I have a look at all the examples, and none of them are using a vector from VectorPostprocessor. I tried defining the user object for the block and then defining 6 functions for each column value, and then defining generic function material to assign 6 

[UserObjects]
  [reader_block]
    type = PropertyReadFile
    prop_file_name = 'Hill_constants.csv'
    read_type = 'block'
    nprop = 6
    nblock = 1
  []
[]

[Functions]
  [bl_1]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '0'
  []
  [bl_2]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '1'
  []
  [bl_3]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '2'
  []
    [bl_4]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '3'
  []
[]

[VectorPostprocessors]
  [vpp]
    type = MaterialVectorPostprocessor
    material = 'mat'
    block = 0
  []
[]


[Materials]
  [mat]
    type = GenericFunctionMaterial
    block = 0
    prop_names = hill_constants
    prop_values = 'bl_1 bl_2 bl_3'
  []
  [hill_tensor]
    type = ADHillConstants
    # F G H L M N
    hill_constants = hill_constants
    use_large_rotation = true
  []
[]


However, I am getting same error
[31m
*** ERROR ***
/storage/work/mvk6158/moose_directories/moosematerialApp/hill_1.i:117.5: cannot convert field 'Materials/hill_tensor/hill_constants' value 'hill_constants' to float
[39m

**Reply by GiudGiud** [reply]
it would have been the right thing to do if the material took a vector of functions as input. It does not.
But modifying it (in C++) to make it do that is not hard so you should consider it

@recuero @SudiptaBiswas
